[SPARK-18838][core] Add separate listener queues to LiveListenerBus.

vanzin opened the pull request:
This change modifies the live listener bus so that all listeners are
added to queues; each queue has its own thread to dispatch events,
making it possible to separate slow listeners from other more
performance-sensitive ones.

The public API has not changed - all listeners added with the existing
"addListener" method, which after this change mostly means all
user-defined listeners, end up in a default queue. Internally, there's
an API allowing listeners to be added to specific queues, and that API
is used to separate the internal Spark listeners into 3 categories:
application status listeners (e.g. UI), executor management (e.g. dynamic
allocation), and the event log.

The queueing logic, while abstracted away in a separate class, is kept
as much as possible hidden away from consumers. Aside from choosing their
queue, there's no code change needed to take advantage of queues.

Test coverage relies on existing tests; a few tests had to be tweaked
because they relied on `LiveListenerBus.postToAll` being synchronous,
and the change makes that method asynchronous. Other tests were simplified
not to use the asynchronous LiveListenerBus.

Commit:
[SPARK-18838][core] Add separate listener queues to LiveListenerBus.

This change modifies the live listener bus so that all listeners are
added to queues; each queue has its own thread to dispatch events,
making it possible to separate slow listeners from other more
performance-sensitive ones.

The public API has not changed - all listeners added with the existing
"addListener" method, which after this change mostly means all
user-defined listeners, end up in a default queue. Internally, there's
an API allowing listeners to be added to specific queues, and that API
is used to separate the internal Spark listeners into 3 categories:
application status listeners (e.g. UI), executor management (e.g. dynamic
allocation), and the event log.

The queueing logic, while abstracted away in a separate class, is kept
as much as possible hidden away from consumers. Aside from choosing their
queue, there's no code change needed to take advantage of queues.

Metrics were also simplified a little bit; the live bus now keeps
track of metrics per queue instead of individual listeners. This is
mostly to make integration with the existing metrics code in `ListenerBus`
easier, without having to refactor the code; that can be done later if
queue-level metrics are not enough.

Test coverage relies on existing tests; a few tests had to be tweaked
because they relied on `LiveListenerBus.postToAll` being synchronous,
and the change makes that method asynchronous. Other tests were simplified
not to use the asynchronous LiveListenerBus.

SparkQA commented:
**[Test build #81693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81693/testReport)** for PR 19211 at commit [`4a6858f`](https://github.com/apache/spark/commit/4a6858f57124a8b405e8ce414f61088a68dc27eb).

SparkQA commented:
**[Test build #81693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81693/testReport)** for PR 19211 at commit [`4a6858f`](https://github.com/apache/spark/commit/4a6858f57124a8b405e8ce414f61088a68dc27eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81693/
Test PASSed.

Commit:
Better name for ListenerQueue.

Commit:
Cleanup metric initialization a bit.

Also fix a problem where the queue size metric might be out of date.

Commit:
Move dispatching to the event thread.

This change makes the event queue implement SparkListenerBus and inherit
all the metrics and dispatching behavior, making the change easier on the
scheduler and also restoring per-listener metrics.

SparkQA commented:
**[Test build #81724 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81724/testReport)** for PR 19211 at commit [`2915a5e`](https://github.com/apache/spark/commit/2915a5ec1bd9d4bc7a40b0ad20ca5b0db8f5382e).

Commit:
Minor cleanup.

SparkQA commented:
**[Test build #81725 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81725/testReport)** for PR 19211 at commit [`cf5c6ce`](https://github.com/apache/spark/commit/cf5c6ce74c185ebd90ea0f9040b177c64161cccc).

Commit:
Undo obsolete changes.

SparkQA commented:
**[Test build #81726 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81726/testReport)** for PR 19211 at commit [`24f5c8d`](https://github.com/apache/spark/commit/24f5c8d0c78a8a362f4690ad03dac9dd07808f85).

Commit:
Fix a comment after changes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81725/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81726/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81724/
Test FAILed.

vanzin commented:
(Nevermind the test failures, I killed the obsolete builds.)

vanzin commented:
retest this please

SparkQA commented:
**[Test build #81728 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81728/testReport)** for PR 19211 at commit [`ad6ff49`](https://github.com/apache/spark/commit/ad6ff49de17c204e8d4feb775185a05d7fa9f53b).

SparkQA commented:
**[Test build #81728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81728/testReport)** for PR 19211 at commit [`ad6ff49`](https://github.com/apache/spark/commit/ad6ff49de17c204e8d4feb775185a05d7fa9f53b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81728/
Test PASSed.

Commit:
Undo unnecessary changes.

Commit:
Make the event log queue name a constant too.

SparkQA commented:
**[Test build #81744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81744/testReport)** for PR 19211 at commit [`20b8382`](https://github.com/apache/spark/commit/20b83826a70ac8574e289db9fdcae37c305c01bd).

SparkQA commented:
**[Test build #81744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81744/testReport)** for PR 19211 at commit [`20b8382`](https://github.com/apache/spark/commit/20b83826a70ac8574e289db9fdcae37c305c01bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81744/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-18838

vanzin commented:
I think this is ready now, pinging some people: @squito @JoshRosen @cloud-fan

SparkQA commented:
**[Test build #81795 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81795/testReport)** for PR 19211 at commit [`1d43d3a`](https://github.com/apache/spark/commit/1d43d3a221fe856cd80052787a11afd8e20419b1).

SparkQA commented:
**[Test build #81795 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81795/testReport)** for PR 19211 at commit [`1d43d3a`](https://github.com/apache/spark/commit/1d43d3a221fe856cd80052787a11afd8e20419b1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

ajbozarth commented:
I did a read-through of your code and didn't find any glaring issues, but I'd like to go through it a bit more before passing judgment. Overall a nice new feature though.

cloud-fan commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -217,7 +217,7 @@ private[spark] class ExecutorAllocationManager(
    */
   def start(): Unit = {
-    listenerBus.addListener(listener)
+    listenerBus.addToQueue(listener, LiveListenerBus.EXECUTOR_MGMT_QUEUE)
```
will we have more queues in the future? If it's not too many, maybe we can use different methods, like `addToMGMTQueue`

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala:
```diff
@@ -0,0 +1,197 @@
+   * Start an asynchronous thread to dispatch events to the underlying listeners.
+   *
+   * @param sc Used to stop the SparkContext in case the a listener fails.
+   * @param metrics Used to report listener performance metrics.
```
there is no `metrics` param.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -65,53 +60,61 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-              processingEvent = false
-            }
-          }
+  private val queues = new CopyOnWriteArrayList[AsyncEventQueue]()
```
can we have something like
```
private val defaultQueue = ...
private val executorManagementQueue = ...
...
```

I think the number of queues will be small.

vanzin replied:
I can add specific methods for each internal queue; but I'd like to keep the internal management of queues more generic. One of the ideas in #18253 was to allow filtering of which events are enqueued at all (e.g. don't enqueue `SparkListenerBlockUpdated` because it's not written to the event logs, reducing the load on the event log queue). Leaving the internal management more generic would allow that to be more easily / cleanly implemented later (instead of "addToEventLogQueue", you'd have a "addCustomQueue" method with a subclass of `AsyncEventQueue`).

Commit:
Explicit add methods for each queue.

Commit:
Rename addListener() to addToSharedQueue().

Commit:
Remove stale comment.

SparkQA commented:
**[Test build #81826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81826/testReport)** for PR 19211 at commit [`ed71477`](https://github.com/apache/spark/commit/ed714778385f8dcb117f3c01a204bd9b024ea83c).

SparkQA commented:
**[Test build #81826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81826/testReport)** for PR 19211 at commit [`ed71477`](https://github.com/apache/spark/commit/ed714778385f8dcb117f3c01a204bd9b024ea83c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81826/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+      case None =>
+        val newQueue = new AsyncEventQueue(queue, conf, metrics)
+        newQueue.addListener(listener)
+        if (started.get() && !stopped.get()) {
```
`stopped.get` is always false here.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+   */
+  private def addToQueue(listener: SparkListenerInterface, queue: String): Unit = synchronized {
+    if (stopped.get()) {
+      throw new IllegalStateException("LiveListenerBus is stopped.")
```
shall we also check `!started.get`?

vanzin replied:
No. It's ok and actually expected to add listeners before the bus is started.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+      }
+      .foreach { toRemove =>
+        if (started.get() && !stopped.get()) {
+          toRemove.stop()
```
this logic is a little weird, why we stop the queue when there is no listeners? is it possible that we may add listeners back to this queue later?

cloud-fan replied:
oh I see your point, we can re-create and start this queue later if needed. Is it really necessary? Can't we keep all the queues active all the time?

vanzin replied:
That would waste resources in the normal case.

The default configuration would never add anything to the default and event log queues, so you'd have two threads dispatching events to no listeners.

Stopping the queue when it empties, while a rare case, handles when users add listeners and later remove them, potentially leaving the default queue empty. Singling out the default queue here just complicates the code.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -65,53 +60,76 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+
+  /** Post an event to all queues. */
+  def post(event: SparkListenerEvent): Unit = {
+    if (!stopped.get()) {
```
shall we add an assert that this bus is started?

vanzin replied:
No, as the class javadoc says, it's ok to send events before the bus is started.

Commit:
Remove redundant check.

SparkQA commented:
**[Test build #81884 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81884/testReport)** for PR 19211 at commit [`77dd8ec`](https://github.com/apache/spark/commit/77dd8ecb549b8f5d7b7b20ac7384b8f56f1df67e).

SparkQA commented:
**[Test build #81884 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81884/testReport)** for PR 19211 at commit [`77dd8ec`](https://github.com/apache/spark/commit/77dd8ecb549b8f5d7b7b20ac7384b8f56f1df67e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81884/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -39,20 +41,13 @@ import org.apache.spark.util.Utils
 
   import LiveListenerBus._
 
   private var sparkContext: SparkContext = _
```
maybe I missed it but I don't see this being set anywhere now

vanzin replied:
Ah, this should be set in `start()`, probably deleted in one of the code moves.

Commit:
Keep reference to SparkContext.

SparkQA commented:
**[Test build #81895 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81895/testReport)** for PR 19211 at commit [`35d3428`](https://github.com/apache/spark/commit/35d3428f6e248c2e21e9190418115fe8c0bbee5d).

SparkQA commented:
**[Test build #81895 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81895/testReport)** for PR 19211 at commit [`35d3428`](https://github.com/apache/spark/commit/35d3428f6e248c2e21e9190418115fe8c0bbee5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81895/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+    queues.asScala.flatMap { queue => queue.findListenersByClass[T]() }
+  }
+
+  private[spark] def listeners: JList[SparkListenerInterface] = {
```
where do we use this method?

vanzin replied:
Used by tests.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-    logTrace(s"Dropping event $event")
   }
+
+  private[spark] def findListenersByClass[T <: SparkListenerInterface : ClassTag](): Seq[T] = {
```
let's add a comment to say it testing only.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -173,80 +159,62 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+
+  private[scheduler] val APP_STATUS_QUEUE = "appStatus"
+
+  private[scheduler] val EXECUTOR_MGMT_QUEUE = "executorMgmt"
```
`MGMT` is hard to understand, how about `MANAGEMENT`?

vanzin replied:
That's a pretty common abbreviation for "management", but sure.

cloud-fan commented on core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala:
```diff
@@ -42,59 +44,65 @@ class SparkListenerSuite extends SparkFunSuite with LocalSparkContext with Match
     assert(counter.count === 5)
-    assert(bus.metrics.numEventsPosted.getCount === 5)
-
-    // Make sure per-listener-class timers were created:
```
do we still have this?

cloud-fan replied:
oh seems we don't need it anymore.

cloud-fan commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -575,8 +575,6 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with TimeL
 
   test("getActive and getActiveOrCreate") {
     require(StreamingContext.getActive().isEmpty, "context exists from before")
-    sc = new SparkContext(conf)
```
why this change?

vanzin replied:
Because the test stops the underlying streaming context and thus the spark context, and you can't add listeners to the bus after that. Thus, it was failing.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -39,20 +41,13 @@ import org.apache.spark.util.Utils
-private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-
-  self =>
+private[spark] class LiveListenerBus(conf: SparkConf) {
```
is it still valid to call `LiveListenerBus.postToAll`? If not, shall we overwrite `postToAll` and throw exception?

cloud-fan replied:
also `LiveListenerBus.addListener` seems not valid too.

vanzin replied:
`LiveListenerBus.postToAll` doesn't exist anymore, and neither does `LiveListenerBus.addListener`. But I'm not sure what you're asking.

cloud-fan replied:
oh sorry I missed that, `LiveListenerBus` doesn't extend `SparkListenerBus` anymore.

cloud-fan commented:
LGTM except some minor comments

jerryshao commented on core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala:
```diff
@@ -0,0 +1,196 @@
+      this.sc = sc
+      dispatchThread.start()
+    } else {
+      throw new IllegalStateException(s"$name already started!")
```
Is it better to be a no-op here, or a warning log instead?

squito replied:
seems fine to me -- its only called in LiveListenerBus, where we guarantee this is true.  seems better to fail-fast if its messed up

Commit:
Nits.

SparkQA commented:
**[Test build #81940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81940/testReport)** for PR 19211 at commit [`283b733`](https://github.com/apache/spark/commit/283b733dbcad326898eee2f4629a8fc1891a77c7).

squito commented:
lgtm

do you know how much better this makes it?  Eg., if we had an existing case where things go haywire in dynamic allocation because of this -- we could see if after this change, the only dropped events are in eventLog, hopefully.  Seems like the right change regardless, but it would be good to know the effect.

vanzin commented:
I don't really have a readily available cluster that can easily put that much pressure on the listener bus; I tried with the one I have and monitored the listener bus metrics, and found pretty much what I expected (slightly more latency in the event log queue, but not enough in my tests to actually cause event drops).

SparkQA commented:
**[Test build #81940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81940/testReport)** for PR 19211 at commit [`283b733`](https://github.com/apache/spark/commit/283b733dbcad326898eee2f4629a8fc1891a77c7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81940/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

dongjoon-hyun commented:
Hi, @vanzin and @cloud-fan .
This seems to break the master branch. Could you take a look this?
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7/3508/

vanzin commented:
Following assert is failing in some cases, let me fix it and send a follow up PR:

```
    val listeners = sc.listenerBus.findListenersByClass[SaveExecutorInfo]
    assert(listeners.size === 1)
```
